### PR TITLE
Reorganized packages, added Util class, Robot observables, and tests

### PR DESCRIPTION
**Pull request by ashergottlieb**

### CODE REVIEW

File: src/com/nutrons/FRamework/InputFactory.java [hunk 1]
```diff
@@ -30,16 +28,12 @@ public static InputFactory instance() {
         return InputFactory.instance;
     }
 
-    public <T> Flowable<T> toFlow(Supplier<T> s, long period, TimeUnit unit) {
-        return Flowable.interval(period, unit).map((x) -> s.get());
-    }
-
     private Joystick lazyJoy(int instance) {
```
Comment (MiloDavis) [inline src/com/nutrons/FRamework/InputFactory.java]: Change name from lazyJoy to memoizedJoy

File: src/com/nutrons/FRamework/InputFactory.java [hunk 2]
```diff
@@ -30,16 +28,12 @@ public static InputFactory instance() {
...
         if (!this.joysticks.containsKey(instance)) {
```
Comment (MiloDavis) [inline src/com/nutrons/FRamework/InputFactory.java]: Rename instance to port to match wpilib doc

File: src/com/nutrons/FRamework/OutputFactory.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    private Map<Integer, SpeedController> controllers;
+    private Map<Integer, Consumer<Double>> motorSpeedConsumer;
+
+    private OutputFactory() {
+        this.controllers = new HashMap<>();
+        this.motorSpeedConsumer = new HashMap<>();
+    }
+
+    public static OutputFactory instance() {
```
Comment (MiloDavis) [inline src/com/nutrons/FRamework/OutputFactory.java]: Make this method synchronized

File: src/com/nutrons/FRamework/OutputFactory.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        return OutputFactory.instance;
+    }
+
+    public void setControllers(Map<Integer, SpeedController> controllers) {
+        this.controllers.clear();
+        this.motorSpeedConsumer.clear();
+        this.controllers.putAll(controllers);
+    }
+
+    private void lazySpeedConsumer(int instance) {
```
Comment (MiloDavis) [inline src/com/nutrons/FRamework/OutputFactory.java]: Rename to memoized

File: src/com/nutrons/FRamework/Robot.java [hunk 5]
```diff
@@ -1,13 +1,39 @@
...
+import static com.nutrons.FRamework.CompMode.*;
 
 public class Robot extends RobotBase {
 
+    private Flowable<Boolean> enabled;
+    private Flowable<CompMode> mode;
+
+    public Robot() {
+        this.enabled = Util.changedValues(Util.toFlow(this::isEnabled));
+        this.mode = Util.changedValues(Flowable.merge(Util.toFlow(this::isAutonomous).filter((x) -> x).map((x) -> AUTO),
```
Comment (MiloDavis) [inline src/com/nutrons/FRamework/Robot.java]: Add some line breaks.  This line is too long

File: src/com/nutrons/FRamework/Robot.java [hunk 6]
```diff
@@ -1,13 +1,39 @@
...
+                Util.toFlow(this::isOperatorControl).filter((x) -> x).map((x) -> TELEOP),
```
Comment (MiloDavis) [inline src/com/nutrons/FRamework/Robot.java]: The identity function is available as a static method in the java.util.Function class.

File: src/com/nutrons/FRamework/Robot.java [hunk 7]
```diff
@@ -1,13 +1,39 @@
...
+                Util.toFlow(this::isTest).filter((x) -> x).map((x) -> TEST)));
```
Comment (MiloDavis) [inline src/com/nutrons/FRamework/Robot.java]: Identity function

File: src/com/nutrons/FRamework/Util.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+public class Util {
+    /**
+     * Generate a Flowable from a periodic call to a Supplier.
+     *
+     * @param period the number of time units to wait before calling the supplier again
+     * @param <T>    the type of the Flowable and Supplier
+     */
+    static <T> Flowable<T> toFlow(Supplier<T> s, long period, TimeUnit unit) {
+        return Flowable.interval(period, unit, Schedulers.io()).subscribeOn(Schedulers.io()).map((x) -> s.get());
```
Comment (MiloDavis) [inline src/com/nutrons/FRamework/Util.java]: Rename parameter to ignored.  Add line break here as well

File: src/com/nutrons/FRamework/Util.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+     */
+    static <T> Flowable<T> toFlow(Supplier<T> s) {
+        return toFlow(s, 100, TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * @return A Flowable of all items that are not equal to their predecessor, including the initial item.
+     */
+    static <T> Flowable<T> changedValues(Flowable<T> o) {
+        Flowable<List<T>> asymmetricPairs = o.buffer(2, 1).filter((x) -> x.size() == 2 && !x.get(0).equals(x.get(1)));
```
Comment (MiloDavis) [inline src/com/nutrons/FRamework/Util.java]: Line breaks

File: src/com/nutrons/FRamework/InputFactory.java [hunk 10]
```diff
@@ -30,16 +28,12 @@ public static InputFactory instance() {
...
             Joystick j = new Joystick(instance);
             this.joysticks.put(instance, j);
-            this.joysticksX.put(instance, this.toFlow(() -> j.getAxis(Joystick.AxisType.kX), 100, TimeUnit.MILLISECONDS));
-            this.joysticksY.put(instance, this.toFlow(() -> j.getAxis(Joystick.AxisType.kY), 100, TimeUnit.MILLISECONDS));
+            this.joysticksX.put(instance, Util.toFlow(() -> j.getAxis(Joystick.AxisType.kX)));
```
Comment (MiloDavis) [inline src/com/nutrons/FRamework/InputFactory.java]: Make a Joystick Observable that returns tuples (x, y) and then add maps to split them

File: src/com/nutrons/FRamework/Util.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    /**
+     * @return A Flowable of all items that are not equal to their predecessor, including the initial item.
+     */
+    static <T> Flowable<T> changedValues(Flowable<T> o) {
+        Flowable<List<T>> asymmetricPairs = o.buffer(2, 1)
+                .filter((x) -> x.size() == 2
+                        && !x.get(0).equals(x.get(1)));
+        return o.take(1).concatWith(asymmetricPairs.map((x) -> x.get(1)));
```
Comment (kreeve) [inline src/com/nutrons/FRamework/Util.java]: Line breaks here too

File: test/com/nutrons/FRamework/test/TestPair.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    @Test
+    public void checkEquality() {
+        assertEquals(p, new Pair<Integer, String>(1, "b"));
+    }
+
+    @Test
+    public void checkInequality() {
+        assertEquals(p, new Pair<Integer, String>(2, "b"));
```
Comment (kreeve) [inline test/com/nutrons/FRamework/test/TestPair.java]: This test fails....

**Review by kreeve — APPROVED** [review]

**Pull request merged**